hatchways/team-penguin
# BE invitation dialog 40

updates create invitation api route (post)

## requirements

at https://github.com/hatchways/team-penguin/issues/40#issuecomment-634453333

## to test

used curl to make BE requests, with the jwt auth step commented out

tested the following cases (verifying results in cloud database with mongo shell; I can provide a connection string):

* invite recipients (`toEmailAr`) includes some users who are registered (c@c.com, d@d.com) and some who are not registered. invitations should be created for registered users. sendgrid should send email invites to the non-registered recipients.

* invite recipients include only registered users (multiple and single)

* invite recipients include only non-registered users (multiple and single)

Notes: I didn't test really hard on the multiples. I'm not entirely clear on the error handling like if there are multiple error scenarios, how to gather all those. I'm aware that there is a bit of redundancy in the code (creating invites and making sendgrid requests but I'm not clear how to make it more dry b/c of the nesting.

curl example:

```
curl -d  '{"toEmailAr": ["your_email1@gmail.com", "your_email2@gmail.com", "c@c.com", "d@d.com"], "fromEmail": "z@z.com", "referralId": "test-referral_id"}' -H "Content-Type: application/json" -X POST http://127.0.0.1:3001/invitations
```

------

(old messages)
@sina-jamshidi I'm a little bit unclear of when I should return a result to the browser. I think the timing is ok for these cases:

* all invitation recipients are registered users

* all invitation recipients are not registered

(update with proposed solution) I suppose I have to check the cases in this order

1 if there are both invite recipients who are registered and who are not registered
   * for this case, I suppose, I might nest the requests like first make the local database changes and after completion, make the sendgrid requests; I will try this

2 if there are invite recipients who are all registered

3 if there are invite recipients who are all not registered
